File: samples/decode_detection_encode/decode_detection_encode.py
#
# Copyright (C) 2025 Vastai-tech Company.
# All rights reserved.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.
#
import os
import sys

current_file_path = os.path.dirname(os.path.abspath(__file__))
common_path = os.path.join(current_file_path, "../..")
sys.path.append(common_path)

from common.detector import Detector
import argparse
import vaststreamx as vsx
import time
import cv2
import os
import numpy as np
import common.utils as utils


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-m",
        "--model_prefix",
        default="/opt/vastai/vaststreamx/data/models/"
        "yolov5m-int8-percentile-1_3_640_640-vacc-pipeline/mod",
        help="model prefix of the model suite files",
    )
    parser.add_argument(
        "--vdsp_params",
        default="./data/configs/yolo_div255_bgr888.json",
        help="vdsp preprocess parameter file",
    )
    parser.add_argument(
        "-d",
        "--device_id",
        default=0,
        type=int,
        help="device id to run",
    )
    parser.add_argument(
        "--threshold",
        default=0.1,
        type=float,
        help="threshold for detection",
    )
    parser.add_argument(
        "--label_file",
        default="data/labels/coco2id.txt",
        help="label file",
    )
    parser.add_argument(
        "--input_uri",
        default="data/videos/test.mp4",
        type=str,
        help="uri to decode",
    )
    parser.add_argument(
        "--output_file",
        default="",
        type=str,
        help="output path",
    )
    args = parser.parse_args()
    return args


if __name__ == "__main__":
    args = argument_parser()
    batch_size = 1
    detector = Detector(args.model_prefix, args.vdsp_params, batch_size, args.device_id)
    detector.set_threshold(args.threshold)
    labels = utils.load_labels(args.label_file)

    cap = vsx.VideoCapture(
        args.input_uri, vsx.CaptureMode.FULLSPEED_MODE, args.device_id, True
    )

    ret, frame, frame_attr = cap.read()
    assert ret

    print(f"Frame width: {frame.width}, height: {frame.height}")
    print(f"Frame rate: {frame_attr.video_fps}")
    print(f"Frame codec_info: {frame_attr.codec_info}")
    print(f"Frame color_space: {frame_attr.color_space}")

    frame_rate = int(frame_attr.video_fps)
    if frame_attr.codec_info.find("avc1") != -1:
        codec_type = vsx.CODEC_TYPE_H264
    elif frame_attr.codec_info.find("hevc") != -1:
        codec_type = vsx.CODEC_TYPE_HEVC
    else:
        print(f"undefined codec_type:{frame_attr.codec_info}")
        exit(-1)

    writer = None
    if args.output_file:
        default_bit_rate = 0
        key_frame_interval = 25
        writer = vsx.VideoWriter(
            args.output_file,
            frame_rate,
            codec_type,
            default_bit_rate,
            key_frame_interval,
            args.device_id,
        )

    while True:
        objects = detector.process(frame)
        bgr_planar = vsx.cvtcolor(frame, vsx.ImageFormat.BGR_PLANAR)
        print(f"bgr_planar width:{bgr_planar.width}")
        print(f"bgr_planar height:{bgr_planar.height}")
        print(f"bgr_planar widthpitch:{bgr_planar.widthpitch}")
        print(f"bgr_planar heightpitch:{bgr_planar.heightpitch}")
        print(f"bgr_planar format:{bgr_planar.format}")
        b_plane, g_plane, r_plane = vsx.as_numpy(bgr_planar)
        bgr_interleave = np.dstack((b_plane, g_plane, r_plane))
        for obj in objects:
            if obj[1] >= 0:
                bbox = [int(obj[2]), int(obj[3]), int(obj[4]), int(obj[5])]
                print(
                    f"Object class: {labels[int(obj[0])]}, score: {obj[1]:.4f}, bbox: {bbox}"
                )
                cv2.rectangle(
                    bgr_interleave,
                    (bbox[0], bbox[1]),
                    (bbox[0] + bbox[2], bbox[1] + bbox[3]),
                    (255, 0, 0),
                    2,
                )
            else:
                break
        yuv_nv12 = utils.cv_bgr888_to_nv12(bgr_interleave)
        vsx_image = vsx.create_image(
            yuv_nv12,
            vsx.ImageFormat.YUV_NV12,
            frame.width,
            frame.height,
            vsx.Context.CPU(),
        )
        if writer:
            writer.write(vsx_image, frame_attr)

        ret, frame, frame_attr = cap.read()
        if ret == 0:
            break
    cap.release()
    if writer:
        writer.release()

File: samples/decode_detection_encode/decode_detection_encode.cpp

/*
 * Copyright (C) 2025 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include <fstream>
#include <iostream>

#include "common/cmdline.hpp"
#include "common/detector.hpp"
#include "opencv2/opencv.hpp"
#include "vaststreamx/vaststreamx.h"

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<std::string>(
      "model_prefix", 'm', "model prefix of the model suite files", false,
      "/opt/vastai/vaststreamx/data/models/"
      "yolov5m-int8-percentile-1_3_640_640-vacc-pipeline/mod");
  args.add<std::string>("hw_config", '\0', "hw-config file of the model suite",
                        false);
  args.add<std::string>("vdsp_params", '\0', "vdsp preprocess parameter file",
                        false, "./data/configs/yolo_div255_yuv_nv12.json");
  args.add<uint32_t>("device_id", 'd', "device id to run", false, 0);
  args.add<float>("threshold", '\0', "threshold for detection", false, 0.5);
  args.add<std::string>("label_file", '\0', "label file", false,
                        "../data/labels/coco2id.txt");
  args.add<std::string>("input_uri", '\0', "input uri", false,
                        "../data/videos/test.mp4");
  args.add<std::string>("output_file", '\0', "output_file", false, "");
  args.parse_check(argc, argv);
  return args;
}

int main(int argc, char* argv[]) {
  // get command parameters
  auto args = ArgumentParser(argc, argv);
  auto device_id = args.get<uint32_t>("device_id");
  // initialize device
  CHECK(vsx::SetDevice(device_id) == 0)
      << "Failed to set device id: " << device_id;
  // initialize model
  auto labels = vsx::LoadLabels(args.get<std::string>("label_file"));
  const int batch_size = 1;
  vsx::Detector detector(args.get<std::string>("model_prefix"),
                         args.get<std::string>("vdsp_params"), batch_size,
                         args.get<uint32_t>("device_id"));
  detector.SetThreshold(args.get<float>("threshold"));

  // open uri
  vsx::VideoCapture cap(args.get<std::string>("input_uri"), vsx::FULLSPEED_MODE,
                        device_id, true, false);
  CHECK(cap.isOpened()) << "Failed to open uri: "
                        << args.get<std::string>("input_uri");
  // output file
  bool save_video = false;
  std::shared_ptr<vsx::VideoWriter> writer;
  auto output_file = args.get<std::string>("output_file");
  if (output_file != "") {
    save_video = true;
    std::cout << "Save video\n";
  }

  vsx::Image frame;  // frame format is nv12, memory is in device
  std::shared_ptr<vsx::FrameAttr> frame_attr;
  if (!cap.read(frame, frame_attr)) {  // get first frame
    std::cout << "Failed to read frame\n";
    return -1;
  }

  // print video info
  std::cout << "Frame width: " << frame.Width()
            << ", width_pitch: " << frame.WidthPitch()
            << ", height: " << frame.Height()
            << ", height_pitch: " << frame.HeightPitch() << std::endl;
  std::cout << "Frame rate: " << frame_attr->video_fps << std::endl;
  std::cout << "Video codec_info: " << frame_attr->codec_info << std::endl;
  if (frame_attr->color_space == vsx::ImageColorSpace::kCOLOR_SPACE_BT709)
    std::cout << "Frame ColorSpace:  BT709 \n";
  else if (frame_attr->color_space == vsx::ImageColorSpace::kCOLOR_SPACE_BT601)
    std::cout << "Frame ColorSpace:  BT601 \n";
  else if (frame_attr->color_space ==
           vsx::ImageColorSpace::kCOLOR_SPACE_BT709_LIMIT_RANGE)
    std::cout << "Frame ColorSpace:  BT709_LIMIT_RANGE \n";
  else if (frame_attr->color_space ==
           vsx::ImageColorSpace::kCOLOR_SPACE_BT601_FULL_RANGE)
    std::cout << "Frame ColorSpace:  BT601_FULL_RANGE \n";

  if (save_video) {
    // int width = frame.Width();
    // int height = frame.Height();
    float frame_rate = frame_attr->video_fps;
    // uint32_t frame_rate_numerator = frame_rate;
    // uint32_t frame_rate_denominator = 1;
    vsx::CodecType codec_type;
    if (frame_attr->codec_info.find("avc1") != std::string::npos) {
      codec_type = vsx::CODEC_TYPE_H264;
    } else if (frame_attr->codec_info.find("hevc") != std::string::npos) {
      codec_type = vsx::CODEC_TYPE_HEVC;
    } else {
      std::cerr << "undefined codec_type:" << frame_attr->codec_info
                << std::endl;
      return -1;
    }

    // open video_writer
    int default_bit_rate = 0;
    int key_frame_interval = 25;
    writer = std::make_shared<vsx::VideoWriter>(
        output_file, static_cast<uint32_t>(frame_rate), codec_type,
        default_bit_rate, key_frame_interval, device_id);
    if (!writer->isOpened()) {
      std::cout << "Failed to open video_writer, params are output_file: "
                << output_file << ", frame_rate: " << frame_rate
                << ", codec_type: " << codec_type
                << ", bit_rate: " << default_bit_rate
                << ", key_frame_interval: " << key_frame_interval
                << ", device_id: " << device_id << std::endl;
      return -1;
    }
  }
  int frame_num = 0;
  do {
    // inference
    auto result = detector.Process(frame);
    auto res_shape = result.Shape();
    const float* res_data = result.Data<float>();
    std::cout << "Frame : " << frame_num++ << " detection objects:\n";
    for (int i = 0; i < res_shape[0]; i++) {
      if (res_data[0] < 0) break;
      std::string class_name = labels[static_cast<int>(res_data[0])];
      float score = res_data[1];
      std::cout << "Object class: " << class_name << ", score: " << score
                << ", bbox: [" << res_data[2] << ", " << res_data[3] << ", "
                << res_data[4] << ", " << res_data[5] << "]\n";
      res_data += vsx::kDetectionOffset;
    }

    if (save_video) {  // draw box in frame
      // cvtcolor to rgb, and copy to cpu
      vsx::Image rgb_image_cpu;
      vsx::CvtColor(frame, rgb_image_cpu, vsx::ImageFormat::RGB_PLANAR,
                    frame_attr->color_space, true);
      // convert to opencv mat
      cv::Mat cv_mat;
      vsx::ConvertVsxImageToCvMatBgrPacked(rgb_image_cpu, cv_mat);
      // draw box
      const float* res_data = result.Data<float>();
      for (int i = 0; i < res_shape[0]; i++) {
        if (res_data[0] < 0) break;
        cv::Rect2f rect = {res_data[2], res_data[3], res_data[4], res_data[5]};
        cv::rectangle(cv_mat, rect, cv::Scalar(0, 255, 0), 2);
        res_data += vsx::kDetectionOffset;
      }
      // convert cv_mat to vsx::Image
      vsx::Image vsx_yuv_image;
      CHECK(vsx::MakeVsxImage(cv_mat, vsx_yuv_image,
                              vsx::ImageFormat::YUV_NV12) == 0);
      // write frame
      writer->write(vsx_yuv_image, frame_attr);
    }
  } while (cap.read(frame, frame_attr));
  std::cout << "Close cap\n";
  cap.release();
  if (save_video) {
    std::cout << "Close writer\n";
    writer->release();
  }
  return 0;
}